### PR TITLE
[FIX] web: dropdown, avoid to scss extend generic classes

### PR DESCRIPTION
**Pull request by stefanorigano**
--
I confirm I have signed the CLA and read the PR guidelines at www.odoo.com/submit-pr

### CODE REVIEW

**Comment by stefanorigano** [comment]
Hello @sdegueldre , thanks again for your workaround.

This fix seems to do the job:
- the generated css weight is ± equal
- css rules are ok (no more dropdown's selectors injected into other components)
- tour tests are green

What's your thoughts about?

**Comment by sdegueldre** [comment]
@stefanorigano It seems like this PR only adresses the ones that I had commented/"fixed" (not really fixed as you may have noticed), but it also looks like some of the `@extend` on classes remain. In general, I would advise completely against using `@extend` on classes, as this effectively makes the sass compiler generate rules so that the selector for the current context is taken into account everywhere else that the class is used. This means that by using `@extend .dropdown-item`, every rule that anyone else writes in any module that targets this class will now be duplicated to also apply to the extending selector.

Also, in my "fix", in an effort to not break the entire design, I only commented the biggest offenders, but there are many more that have a significant impact on compilation time (currently the bundle generation takes ~2s and there are some rules that take up .5s of that by themselves for a single `@extend`). If possible, it would be good if you could completely remove any `@extend` rule that applies to a class, and use placeholders instead (like you're doing here). I think we can bring these compilation times down to a much more reasonable level, as 2 seconds for compiling just the scss is still a very long time to make users wait.

**Comment by stefanorigano** [comment]
@sdegueldre I've initially misunderstood the amount of changes you were suggesting...
We can't just replace the `@extend` with few css lines, removing `@extend`for classes requires a bit of assets reorganisation.

Let me explain why (and check if everyone is on the same page).
We want an architecture that ensures:

1. Simple maintenance.
2. Visual consistency between Owl, jQueryUI and Bootstrap dropdowns.
3. Backport compatibility for modules legacy code.
4. Avoiding js conflict (eg. Bootstrap js events fired on  Owl components).

`@extend` was the obvious choose because it allows to "_just apply Bootstrap design_" to Owl and jQueryUI components achieving all 4 objectives, indeed when you said:

> This means that by using @extend .dropdown-item, every rule that anyone else writes in any module that targets this class will now be duplicated to also apply to the extending selector.

you're right, and it's actually intentional. Indeed when a module adds a custom rules for `.dropdown-item` we want these rules to affect Owl `.o_dropdown_item` too, since both should behave and look consistent in that context (point 3).

But as you explained performance downsides are serious, indeed I may have underestimated the amount of customisations modules apply on these components.

To remove the `@extend`s I propose 2 different approaches:

1. Copy/Paste all bootstrap code ( [~190 lines](https://github.com/twbs/bootstrap/blob/v4-dev/scss/_dropdown.scss)) and apply it to Owl and jQueryUI components. It will ensure visual consistency (👍 ) but we'll lose backport compatibility (👎 ) and we'll have to maintain 3 different versions of the same design (👎 ).

2. Add Boostrap classes directly into Owl and jQueryUI components (...yes I know but... why not?) .

Solution 2 seems feasible to me. I pushed https://github.com/odoo/odoo/pull/75302/commits/28e0bda1b43c2f9228540725240aa6e1ff56a7da as a POC.
Beside adding Bootstrap classes, this commit adapt Owl xml to mirror Bootstrap, replacing the `ul > li > [whatever]` structure with `div > [whatever]`.
This last change is necessary for the POC but I'd like to propose it anyways 'cause it removes the need of a `t-tag`, deeply simplify html/css and ease `<form>` injection.
Points 1, 2 and 3 are achieved (👍 👍 👍 ), but I'm not sure about point 4 (🤷🏻‍♂️) ... in my test everything seems fine, but I may miss something.

As said it's an huge change, @ged-odoo  and @aab-odoo should be involved in the discussion too (probably @antonylesuisse too since he was in touch with Gery about this issue).

Please let me know how to proceed, thanks!

cc. @brboi

**Review by brboi — COMMENTED** [review]
@stefanorigano I agree with your POC

IMO it is a bit sad to kind of duplicate css classes (not a huge fan of seeing i.e. "o_dropdown_item dropdown-item" everywhere) but I do agree that this solution works.
Maybe we just need to find better usage/naming of our "o_..." classes, but this could be done later on.

I left some comments here and there.

here are some more:
- the dropdown keyboard navigation is broken
- the navbar sections are `<ul/>`s and now that `<Dropdown/>`s are `<div/>`s the final DOM is not compliant with HTML specs. (ul element should only contain li elements)

I also tested your poc in master-wowl and added a commit on top.
See https://github.com/odoo-dev/odoo/compare/master-wowl...odoo-dev:master-wowl-dropdown-extends-sri-boi

cc @sdegueldre

File: addons/web/static/src/core/dropdown/dropdown.scss [hunk 1]
```diff
@@ -52,83 +47,39 @@
 
 .o_rtl {
   .dropdown-menu {
-    @extend .dropdown-menu-right;
+    // @extend .dropdown-menu-right;
```
[the remark this answers is not in the document]
Comment (brboi) [reply addons/web/static/src/core/dropdown/dropdown.scss:50]: done in https://github.com/odoo/odoo/pull/76389

**Comment by stefanorigano** [comment]
Hello @qsm-odoo ,

I know that is difficult to find some time on these days, but just an opinion would be very appreciated :)
We want to "extend" bootstrap dropdown's design  to both Owl and jQuery UI components, but without using sass `@extend` feature for performance reason.

I think that the only feasible solution is to apply bootstrap classes directly in the components (my analysis [here](https://github.com/odoo/odoo/pull/75302#issuecomment-903232046) ).

What's your thoughts?

**Comment by qsm-odoo** [comment]
@stefanorigano I agree and don't see why that would be considered a bad solution at all without reading the whole debate here. That's literally how bootstrap is supposed to be used: add the classes at the right place. We don't like that? Then we should not use bootstrap :wink: (and I don't want that, still dreaming of a time where backend components like views could be added on the website and bam: we have a well-designed-theme-dependant view :open_mouth:)

Even if the `@extend` solution did work without performance problems... I would not have said it is good (why doubling the code and the scss size while you could just use the classes at the right place in the XML).

**Comment by stefanorigano** [comment]
Thanks @qsm-odoo 👍

**Pull request closed without merging**